### DISCUSSION TITLE
cannot set up more than one connection

### DISCUSSION THREAD

**Discussion by HK-13** [alibaba/GraphScope#1975] [General]
目前仅支持建立单个session，建立新session必须先close
但实际情况中，如果多个用户并发需要建立多个session该如何解决？

>比如：
>A用户建立session，加载数据，计算等等操作，此时还无法关闭session
>B用户此时也需要发起session，加载其他得数据，此时就会报错

![image](https://user-images.githubusercontent.com/5295924/185068151-8f5be7ed-8897-43d2-84c5-4bc5dfd25c5e.png)

**Comment by TerryWang1208** [comment]
我们也遇到了同样的问题。所以也想请教一下graphscoep团队，目前是否只能考虑使用多instance来对应多请求处理，每个instance处理一个请求，对instance的创建销毁做好管理。

**Reply by HK-13** [reply]
请教下，所指的单个instance处理多个请求，该如何实现。您这边实现了吗？
单个session在多个请求间共享？

**Reply by sighingnow** [reply]
> 所指的单个instance处理多个请求，该如何实现。您这边实现了吗？

@4613，您是希望一个session，可能从多个线程中去run，还是让多个用户在多个python进程中去用？如果是后者的话，目前的设计和实现不支持，目前的roadmap里也没有这样的考虑哈。

**Reply by HK-13** [reply]
我昨天查到session的reconnect参数，可以复用session。文档中比较委婉的表达使用此特性的可能问题。
能否回复下，如果我打算使用此特性，可能会造成一些比较具体的问题，比如：
1. 复用太多是否会造成资源不可控，
2. 或者，复用后内部加载不同的图数据是否会造成其他问题
3. 等等

**Comment by sighingnow** [comment]
出于简化状态维护的逻辑考虑，每个graphscope的cluster在同一个时刻只允许一个session。

如果需要同时有多个session，可以拉起多个cluster。

/cc @4613 @TerryWang1208

**Reply by TerryWang1208** [reply]
感谢解答~  请问roadmap里是否有考虑增加对同一时刻多session的支持？

**Reply by sighingnow** [reply]
目前不考虑支持。对于需要多个session的情形，还是建议拉起多个cluster。

**Reply by HK-13** [reply]
我一直以为是我用的不对
@sighingnow 仅仅是为了简化逻辑考虑吗？多cluster相对其他计算框架个人感觉比较不合理啊。
在面向生产级的很多应用不都是并发吗？graphscope的目标难道不包含此类场景，或者我的问题是从这个问题出发graphscope的受众场景应该是怎样的？

**Comment by sighingnow** [comment]
hi @4613 @TerryWang1208 ,

> 每个instance处理一个请求，对instance的创建销毁做好管理。

不是的，graphscope的设计是希望用户用一个session去完成一个完整的图计算的workflow，例如我们在 [README] (https://github.com/alibaba/GraphScope) 里面给出的例子，而不是说每个请求一个instance，拉一次instance对应的是一个完整的图计算的job。类似比spark，一个job是指提交一个jar，这个jar里面可能有很多map、reduce的steps。

GraphScope现在release的版本也不针对作为long-running的服务部署的场景（未来会有对应的模块来支持），不是一个graph database。

> 多cluster相对其他计算框架个人感觉比较不合理啊。

graphscope不是一个shared serive，就例如你跑一个tensorflow job，也是大家各自起各自的ps server，ps worker。

> graphscope的目标难道不包含此类场景，或者我的问题是从这个问题出发graphscope的受众场景应该是怎样的？

一个graphscope cluster 应该被用户一个graph computing的job。
